staging/src/k8s.io/kubelet/pkg/cri/streaming/portforward/httpstream.go

```diff
@@ -242,8 +243,16 @@ Loop:
 // function for the given stream pair.
 func (h *httpStreamHandler) portForward(p *httpStreamPair) {
 	ctx := context.Background()
-	defer p.dataStream.Close()
-	defer p.errorStream.Close()
+	resetStreams := false
+	defer func() {
+		if resetStreams {
+			p.dataStream.Reset()  //nolint: errcheck
+			p.errorStream.Reset() //nolint: errcheck
+			return
```

Comment:
What does calling Reset() do, compared to calling Close()?

I had to page back in a little of the Reset() implementation, at first I thought this was trying to leave streams in a state where they could be retried or resumed. Is the intent to more completely / aggressively / immediately tear down and close the dataStream / errorStream?

Reply:
great question!

I have some explanations here, if you still have questions, please let me know. 😄

https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-1543873482

Reply:
I do have a bunch of questions 😄 Happy to leave them here or meet to talk through them.

1. My understanding is that calling Close() is leaving things in a hung state, and the desire here by calling Reset() is to tear the connection down more aggressively / completely. Is that correct?

2. Is the Close() call itself blocking, or just leaving the stream or connection in a hung state?

3. Is there any benefit to calling Close() _and_ Reset() in this case? The docs inside resetStream appear to support this:

   ```go
   func (s *Stream) resetStream() error {
   	// Always call closeRemoteChannels, even if s.finished is already true.
   	// This makes it so that stream.Close() followed by stream.Reset() allows
   	// stream.Read() to unblock.
   ```

4. Both Close() and Reset() attempt to write data to the connection.

   Close():
   ```go
   	return s.WriteData([]byte{}, true)
   ```
   
   Reset() → resetStream():
   ```go
   	return s.conn.framer.WriteFrame(resetFrame)
   ```

   Is that write blocking? If so, should we set deadlines on the stream before calling Close() / Reset()?

Reply:
> Your question is great, I need to do some exploring.

I made some new discoveries while doing testing today, and I hope I can give some clues and conclusions tomorrow.

Reply:
The clues I have now are as follows (please correct me if I'm wrong):

kubectl port-forward will complete the establishment of a TCP connection with the remote end [kubectl ⇒ kube-apiserver ⇒ kubelet ⇒ containerd] the moment it starts ForwardPorts. Subsequent Connections received from the local Socket will be reused on this connection [just create **Stream** instead of **Conn**]

- https://github.com/kubernetes/kubernetes/blob/a9f04103854893056237a09250ad3335867b0391/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L345
- https://github.com/kubernetes/kubernetes/blob/a9f04103854893056237a09250ad3335867b0391/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L194

Reply:
As shown in the figure, 8080 (server) has started the FIN process with containerd (client), so io.EOF will be received at [io.Copy(stream, conn)](https://github.com/containerd/containerd/blob/5f440cbb79f60e7cda1978a83837771ba0a1574b/internal/cri/server/sandbox_portforward_linux.go#L86), but it is masked as nil by io.Copy

<img width="1874" alt="pcap" src="https://github.com/kubernetes/kubernetes/assets/11829907/cb390325-baa1-48e4-8cc4-1b53d3053eea">

Reply:
> My previous troubleshooting may have misled you. 😅

In fact, httpStream#Close() or httpStream#Reset() will not affect the **Coon**. The **Conn** needs Conn#Close() to close and unblock https://github.com/containerd /containerd/blob/03d81f595f4e0c1323591ea48275e96830915502/vendor/k8s.io/kubelet/pkg/cri/streaming/portforward/httpstream.go#L218

Reply:
Regarding your questions:

> My understanding is that calling Close() is leaving things in a hung state, and the desire here by calling Reset() is to tear the connection down more aggressively / completely. Is that correct?

httpStream#Close() will send Fin to the peer, but the local Stream cleaning work depends on the peer's reply to Fin (https://github.com/containerd/containerd/blob/e634f04d8cdb6c2f96eea2e4e66d0e4500a46282/vendor/github.com/moby/spdystream/ stream.go#L173)

> Is the Close() call itself blocking, or just leaving the stream or connection in a hung state?
     
httpStream#Close() will not block, but will put the stream in a huang state. After receiving the peer sending Fin from the httpStream, the local Stream will be cleaned up.

- https://github.com/containerd/containerd/blob/e634f04d8cdb6c2f96eea2e4e66d0e4500a46282/vendor/github.com/moby/spdystream/stream.go#L173


> Is there any benefit to calling Close() *and* Reset() in this case? The docs inside resetStream appear to support this: Is that write blocking? If so, should we set deadlines on the stream before calling Close() /Reset()?

SetDeadlines is meaningful, whether it is httpStream#Close() or httpStream#Reset(), the data packet will be sent.

> Both Close() and Reset() attempt to write data to the connection.Is that write blocking? If so, should we set deadlines on the stream before calling Close() / Reset()?

Neither httpStream#Close() nor httpStream#Reset() will block

Reply:
I hope I can answer your questions. If you still have questions, please let me know. 😄

Reply:
@liggitt PTAL ~ thx

Reply:
>> Is there any benefit to calling Close() and Reset() in this case? The docs inside resetStream appear to support this:
>
> I think httpStream#Reset() is enough

Is Close() more graceful than Reset()? Is there a benefit to attempting the graceful close first, and then following it with the hard close of the stream?

Reply:
httpStream#Close() will send FIN to the peer. [You need to rely on the peer to send FIN (that is, the client calls httpStream#Close()) to clear the local stream] ref: https://www.chromium.org/spdy/spdy-protocol/spdy-protocol-draft3-1/#236-stream-half-close

httpStream#Reset() will remove the local stream and send RST

Because this is two-way traffic, the approach based on httpStream#Close() may be more graceful, but we need to decide how long the interval between Close and Reset is.